File: src/app/modelPresets.ts
export type ModelPresetProviderKind = "openrouter" | "openai_compatible" | "custom";
export type CostTier = "free" | "low" | "standard" | "premium";
export type LatencyTier = "fast" | "balanced" | "quality";

export type ModelPreset = {
  id: string;
  title: string;
  providerKind: ModelPresetProviderKind;
  baseUrl: string;
  primaryModel: string;
  fallbackModels: string[];
  costTier: CostTier;
  latencyTier: LatencyTier;
  supportsStructuredOutputs: boolean;
  requiresCredits: boolean;
  freeTierCaveats: string;
  lastReviewedAt: string;
  notes: string;
};

export const MODEL_PRESETS: readonly ModelPreset[] = [
  {
    id: "openrouter_free_dev",
    title: "OpenRouter Free / Dev",
    providerKind: "openrouter",
    baseUrl: "https://openrouter.ai/api/v1",
    primaryModel: "openai/gpt-4o-mini",
    fallbackModels: ["google/gemini-2.0-flash-exp:free", "meta-llama/llama-3.3-70b-instruct:free"],
    costTier: "free",
    latencyTier: "fast",
    supportsStructuredOutputs: true,
    requiresCredits: false,
    freeTierCaveats: "Free models can be rate-limited and availability may vary.",
    lastReviewedAt: "2026-05-17",
    notes: "Good for development checks. Validate output quality before production use.",
  },
  {
    id: "openrouter_fast_budget",
    title: "OpenRouter Fast / Budget",
    providerKind: "openrouter",
    baseUrl: "https://openrouter.ai/api/v1",
    primaryModel: "openai/gpt-4o-mini",
    fallbackModels: ["anthropic/claude-3.5-haiku", "google/gemini-2.0-flash-001"],
    costTier: "low",
    latencyTier: "fast",
    supportsStructuredOutputs: true,
    requiresCredits: true,
    freeTierCaveats: "",
    lastReviewedAt: "2026-05-17",
    notes: "Lower cost profile for steady throughput.",
  },
  {
    id: "openrouter_balanced_paid",
    title: "OpenRouter Balanced Paid",
    providerKind: "openrouter",
    baseUrl: "https://openrouter.ai/api/v1",
    primaryModel: "openai/gpt-4.1-mini",
    fallbackModels: ["anthropic/claude-3.7-sonnet", "google/gemini-2.5-flash"],
    costTier: "standard",
    latencyTier: "balanced",
    supportsStructuredOutputs: true,
    requiresCredits: true,
    freeTierCaveats: "",
    lastReviewedAt: "2026-05-17",
    notes: "Balanced quality/cost for daily interview sessions.",
  },
  {
    id: "openrouter_quality_paid",
    title: "OpenRouter Quality Paid",
    providerKind: "openrouter",
    baseUrl: "https://openrouter.ai/api/v1",
    primaryModel: "anthropic/claude-3.7-sonnet",
    fallbackModels: ["openai/gpt-4.1", "google/gemini-2.5-pro"],
    costTier: "premium",
    latencyTier: "quality",
    supportsStructuredOutputs: true,
    requiresCredits: true,
    freeTierCaveats: "",
    lastReviewedAt: "2026-05-17",
    notes: "Higher quality profile. Validate against golden dataset before rollout.",
  },
  {
    id: "custom_openai_compatible",
    title: "Custom OpenAI-compatible",
    providerKind: "custom",
    baseUrl: "",
    primaryModel: "",
    fallbackModels: [],
    costTier: "standard",
    latencyTier: "balanced",
    supportsStructuredOutputs: true,
    requiresCredits: false,
    freeTierCaveats: "",
    lastReviewedAt: "2026-05-17",
    notes: "Manual route. Use your own base URL and model.",
  },
] as const;

export const DEFAULT_MODEL_PRESET_ID = "custom_openai_compatible";

export function resolveModelPreset(presetId: string): ModelPreset {
  return (
    MODEL_PRESETS.find((preset) => preset.id === presetId) ??
    MODEL_PRESETS.find((preset) => preset.id === DEFAULT_MODEL_PRESET_ID)!
  );
}

File: src-tauri/src/model_presets.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProviderKind {
    OpenRouter,
    OpenAiCompatible,
    Custom,
}

#[derive(Debug, Clone)]
pub struct ModelPreset {
    pub provider_kind: ProviderKind,
    pub fallback_models: &'static [&'static str],
}

const OPENROUTER_FREE_FALLBACKS: &[&str] = &[
    "google/gemini-2.0-flash-exp:free",
    "meta-llama/llama-3.3-70b-instruct:free",
];
const OPENROUTER_FAST_FALLBACKS: &[&str] =
    &["anthropic/claude-3.5-haiku", "google/gemini-2.0-flash-001"];
const OPENROUTER_BALANCED_FALLBACKS: &[&str] =
    &["anthropic/claude-3.7-sonnet", "google/gemini-2.5-flash"];
const OPENROUTER_QUALITY_FALLBACKS: &[&str] = &["openai/gpt-4.1", "google/gemini-2.5-pro"];

pub fn resolve_model_preset(id: &str) -> ModelPreset {
    match id.trim() {
        "openrouter_free_dev" => ModelPreset {
            provider_kind: ProviderKind::OpenRouter,
            fallback_models: OPENROUTER_FREE_FALLBACKS,
        },
        "openrouter_fast_budget" => ModelPreset {
            provider_kind: ProviderKind::OpenRouter,
            fallback_models: OPENROUTER_FAST_FALLBACKS,
        },
        "openrouter_balanced_paid" => ModelPreset {
            provider_kind: ProviderKind::OpenRouter,
            fallback_models: OPENROUTER_BALANCED_FALLBACKS,
        },
        "openrouter_quality_paid" => ModelPreset {
            provider_kind: ProviderKind::OpenRouter,
            fallback_models: OPENROUTER_QUALITY_FALLBACKS,
        },
        "custom_openai_compatible" => ModelPreset {
            provider_kind: ProviderKind::Custom,
            fallback_models: &[],
        },
        _ => ModelPreset {
            provider_kind: ProviderKind::OpenAiCompatible,
            fallback_models: &[],
        },
    }
}

#[cfg(test)]
mod tests {
    use super::{resolve_model_preset, ProviderKind};

    #[test]
    fn openrouter_preset_contains_fallback_models() {
        let preset = resolve_model_preset("openrouter_free_dev");
        assert_eq!(preset.provider_kind, ProviderKind::OpenRouter);
        assert!(!preset.fallback_models.is_empty());
    }

    #[test]
    fn custom_preset_has_no_fallback_models() {
        let preset = resolve_model_preset("custom_openai_compatible");
        assert_eq!(preset.provider_kind, ProviderKind::Custom);
        assert!(preset.fallback_models.is_empty());
    }
}
